refactor: move examples to common location

yellowred opened the pull request:
## Description

Refactor examples into separate binary crates.

Fixes https://github.com/libp2p/rust-libp2p/issues/3111.

## Links to any relevant issues

## Change checklist

<!-- Please add a Changelog entry in the appropriate crates and bump the crate versions if needed. See <https://github.com/libp2p/rust-libp2p/blob/master/docs/release.md#development-between-releases>-->

- [x] I have performed a self-review of my own code
- [x] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] A changelog entry has been made in the appropriate crates

yellowred committed:
Moved the Chat example into example crate.

yellowred committed:
Remove unnecessary deps

thomaseizinger commented on examples/chat/Cargo.toml:
```diff
@@ -0,0 +1,24 @@
+name = "chat"
+version = "0.1.0"
+edition = "2021"
+
```
Let's set this one to `publish = false` please.

thomaseizinger commented on examples/chat/Cargo.toml:
```diff
@@ -0,0 +1,24 @@
+    "macros",
+    "tcp",
+    "websocket",
+    "yamux"] }
```
Can we format this to a single line please?

thomaseizinger commented on examples/README.md:
```diff
@@ -4,56 +4,7 @@ A set of examples showcasing how to use rust-libp2p.
-  While obviously showcasing how to build a basic file sharing application with the Kademlia and
-  Request-Response protocol, the actual goal of this example is **to show how to integrate
-  rust-libp2p into a larger application**.
+- [Chat](./chat)
```
I think a single head-line would still be good.

thomaseizinger commented on examples/README.md:
```diff
@@ -4,56 +4,7 @@ A set of examples showcasing how to use rust-libp2p.
 
-- [Chat](./chat.rs)
-
-   A basic chat application demonstrating libp2p and the mDNS and floodsub protocols.
-
-    - [Gossipsub chat](./gossipsub-chat.rs)
-
-      Same as the chat example but using mDNS and the Gossipsub protocol.
-
-    - [Tokio based chat](./chat-tokio.rs)
-
-      Same as the chat example but using tokio for all asynchronous tasks and I/O.
```
One thing to tidy up here could be:

- Replace current chat example with gossipsub. It is preferred over floodsub.
- Remove the tokio chat example. I'd rather just convert one of the others to use tokio instead of having a dedicated clone of an example just with another executor.

thomaseizinger reviewed (commented):
Excellent start. I am pleased that Git recognizes these as "moved". That should make this fairly easy to review and merge.

yellowred committed:
Replace chat example with gossipsub version.

yellowred committed:
Moving more examples to separate binary crates.

thomaseizinger commented on examples/ipfs-kad/Cargo.toml:
```diff
@@ -0,0 +1,13 @@
+env_logger = "0.10"
+futures = "0.3.26"
+libp2p = { path = "../../", features=["async-std", "dns", "kad",  "mplex", "noise", "tcp", "websocket", "yamux"] }
+multiaddr = { version = "0.17.0" }
```
Can we add these newlines at the end please?

thomaseizinger replied:
We should be able to import `multiaddr` through `libp2p::core::multiaddr`. Why is the separate dependency needed?

yellowred committed:
Move Rendezvous protocol examples into separate binary crates.

thomaseizinger commented on examples/rendezvous-register/Cargo.toml:
```diff
@@ -0,0 +1,15 @@
+[package]
+name = "rendezvous-register"
```
I think for rendezvous, it makes sense to have one crate "rendezvous" with three binaries as `src/bin/{register,server,discover}.rs`.

yellowred committed:
Gather Rendezvous examples in a single crate.

yellowred committed:
Move a DCUtR example into a separate crate.

yellowred committed:
No publishing for dcutr example

yellowred committed:
Move Identify example to a separate crate.

yellowred committed:
Added new lines into Cargo.toml files in examples.

yellowred committed:
cleanup

yellowred committed:
Fix CI error:
```
error: document output filename collision
The bin `ping` in package `ping v0.1.0 (/home/runner/work/rust-libp2p/rust-libp2p/examples/ping)` has the same name as the bin `ping` in package `interop-tests v0.1.0 (/home/runner/work/rust-libp2p/rust-libp2p/interop-tests)`.
```

yellowred committed:
Fix IPFS integration tests.

yellowred committed:
Use namesapced binaries to avoid name collision in CI.

thomaseizinger commented on Cargo.toml:
```diff
@@ -180,42 +189,6 @@ members = [
     "interop-tests"
 ]
 
-[[example]]
-name = "chat"
-required-features = ["full"]
-
-[[example]]
-name = "chat-tokio"
-required-features = ["full"]
-
-[[example]]
-name = "file-sharing"
-required-features = ["full"]
-
-[[example]]
-name = "gossipsub-chat"
-required-features = ["full"]
-
-[[example]]
-name = "ipfs-private"
-required-features = ["full"]
-
-[[example]]
-name = "ipfs-kad"
-required-features = ["full"]
-
-[[example]]
-name = "ping"
-required-features = ["full"]
-
-[[example]]
-name = "mdns-passive-discovery"
-required-features = ["full"]
-
-[[example]]
-name = "distributed-key-value-store"
-required-features = ["full"]
```
:tada:

thomaseizinger commented on examples/dcutr/Cargo.toml:
```diff
@@ -0,0 +1,13 @@
+env_logger = "0.10.0"
+futures = "0.3.26"
+futures-timer = "3.0"
+libp2p = { path = "../../", features=["async-std", "dns", "dcutr", "identify", "macros", "mplex", "noise", "ping", "relay", "rendezvous", "tcp", "tokio", "yamux"] }
```
```suggestion
libp2p = { path = "../../", features = ["async-std", "dns", "dcutr", "identify", "macros", "mplex", "noise", "ping", "relay", "rendezvous", "tcp", "tokio", "yamux"] }
```

Formatting nit.

thomaseizinger commented on examples/dcutr/src/main.rs:
```diff
@@ -101,15 +102,11 @@ fn main() -> Result<(), Box<dyn Error>> {
-        event_process = false,
-        prelude = "libp2p_swarm::derive_prelude"
-    )]
+    #[behaviour(out_event = "Event", event_process = false, prelude = "derive_prelude")]
```
```suggestion
    #[behaviour(out_event = "Event", event_process = false)]
```

When depending on the `libp2p` crate, this is not needed!

thomaseizinger commented on examples/rendezvous/Cargo.toml:
```diff
@@ -0,0 +1,27 @@
+
+[[bin]]
+name = "rzv-discover"
+path = "src/bin/discover.rs"
```
I'd be in favor of just renaming the files and rely on auto-discover of the binaries. Convention over configuration.

thomaseizinger commented on examples/rendezvous/src/bin/register.rs:
```diff
@@ -131,13 +131,20 @@ impl From<ping::Event> for MyEvent {
     out_event = "MyEvent",
     event_process = false,
-    prelude = "libp2p_swarm::derive_prelude"
+    prelude = "derive_prelude"
```
In all of these, it should be possible to remove this now!

thomaseizinger commented on examples/rendezvous/src/bin/register.rs:
```diff
@@ -131,13 +131,20 @@ impl From<ping::Event> for MyEvent {
 #[derive(NetworkBehaviour)]
 #[behaviour(
     out_event = "MyEvent",
     event_process = false,
```
That has been the default for a while as well, can you remove this too please?

thomaseizinger commented on examples/rendezvous/src/main.rs:
```diff
@@ -115,18 +124,18 @@ impl From<rendezvous::server::Event> for MyEvent {
     }
 }
 
-impl From<ping::Event> for MyEvent {
-    fn from(event: ping::Event) -> Self {
-        MyEvent::Ping(event)
-    }
-}
```
Can we undo this reordering of code please?

thomaseizinger commented on examples/rendezvous/src/main.rs:
```diff
@@ -103,10 +111,11 @@ async fn main() {
+#[allow(clippy::large_enum_variant)]
 enum MyEvent {
     Rendezvous(rendezvous::server::Event),
-    Ping(ping::Event),
```
Here too.

thomaseizinger commented on examples/rendezvous/src/main.rs:
```diff
@@ -18,53 +18,61 @@
-    swarm
-        .listen_on("/ip4/0.0.0.0/tcp/62649".parse().unwrap())
-        .unwrap();
+    let _ = swarm.listen_on("/ip4/0.0.0.0/tcp/0".parse().unwrap());
```
Why do we ignore this error now instead of unwrapping?

thomaseizinger commented on examples/rendezvous/src/bin/register.rs:
```diff
@@ -57,9 +59,7 @@ async fn main() {
 
-    let _ = swarm.listen_on("/ip4/0.0.0.0/tcp/0".parse().unwrap());
-
-    swarm.dial(rendezvous_point_address).unwrap();
```
Why does this example need changing?

thomaseizinger commented on examples/rendezvous/src/bin/identify.rs:
```diff
@@ -19,48 +19,48 @@
     let _ = swarm.listen_on("/ip4/0.0.0.0/tcp/0".parse().unwrap());
 
-    swarm.dial(rendezvous_point_address).unwrap();
+    swarm.dial(rendezvous_point_address.clone()).unwrap();
```
Why do we suddenly need to clone this?

thomaseizinger commented on examples/chat/Cargo.toml:
```diff
@@ -0,0 +1,12 @@
+[package]
+name = "chat"
```
Would it make sense to append `example` to these like `chat-example`? Then it would show up as "Building chat-example" in the cargo logs which I'd mildly prefer.

thomaseizinger commented on examples/rendezvous/src/bin/register.rs:
```diff
@@ -19,35 +19,37 @@
-            .authenticate(libp2p_noise::NoiseAuthenticated::xx(&identity).unwrap())
-            .multiplex(libp2p_yamux::YamuxConfig::default())
+            .authenticate(libp2p::noise::NoiseAuthenticated::xx(&key_pair).unwrap())
+            .multiplex(libp2p::yamux::YamuxConfig::default())
```
We are [in the process](https://github.com/libp2p/rust-libp2p/issues/2217) of apply more consistent naming to all our crates to make usage easier and clearer. Can we import the submodules here please like `use libp2p::tcp;` and then say `tcp::tokio::Transport`?

yellowred committed:
Fix the docs markup

thomaseizinger commented on examples/rendezvous/src/bin/register.rs:
```diff
@@ -131,13 +131,20 @@ impl From<ping::Event> for MyEvent {
+
 #[derive(NetworkBehaviour)]
 #[behaviour(
     out_event = "MyEvent",
```
In a recent version of the `NetworkBehaviour` derive, we add the ability to auto-generate this enum. I think our examples should use that as we consider it the recommended way of using the derive. Can you adapt the examples please? The only thing that will needed changing is that the generated enum will be called `MyBehaviourEvent`.

thomaseizinger commented on examples/pingpong/Cargo.toml:
```diff
@@ -0,0 +1,12 @@
+[package]
+name = "pingpong"
```
```suggestion
name = "ping-example"
```

thomaseizinger commented on examples/file-sharing/src/main.rs:
```diff
@@ -45,7 +45,7 @@
 //! ## Architectural properties
 //!
-//! - Clean clonable async/await interface ([`Client`]) to interact with the
+//! - Clean clonable async/await interface ([`Client`](network::Client)) to interact with the
```
If you wanted, you could split this inner `network` module now into a separate file.

thomaseizinger reviewed (commented):
Thanks for pushing this forward!

I made some comments, most of them a pretty minor. I really like the direction of this :)

yellowred committed:
Improved naming and formatting

yellowred committed:
Change to relative references for easier refactoring later

yellowred committed:
Use an Event derived by macro rather than defining it myself.

yellowred committed:
More improvements in naming

yellowred committed:
Split network module into a separate file.

thomaseizinger commented on examples/rendezvous/Cargo.toml:
```diff
@@ -0,0 +1,23 @@
+name = "rzv-register"
+
+[[bin]]
+name = "rzv-identify"
```
```suggestion
```

This is obsolete now I think.

thomaseizinger commented on examples/rendezvous/src/bin/rzv-identify.rs:
```diff
@@ -19,48 +19,47 @@
-
     let mut swarm = Swarm::with_tokio_executor(
-        libp2p_tcp::tokio::Transport::default()
+        libp2p::tcp::tokio::Transport::default()
```
```suggestion
        tcp::tokio::Transport::default()
```

Nit: Could we also align these please? i.e. add `use libp2p::tcp`.

thomaseizinger commented on examples/rendezvous/src/bin/rzv-identify.rs:
```diff
@@ -19,48 +19,47 @@
         .parse()
         .unwrap();
 
-    let identity = identity::Keypair::generate_ed25519();
```
Did we have to move and rename that? In general I'd like to keep diffs as small as possible.

yellowred replied:
No, we don't, but it is a part of the unification of the Rendezvous examples, makes it clear what is the same and what is different in them. Previously keypair was generated in a random line in the code, now it sticks to the beginning like in any other example.

thomaseizinger replied:
Got it, thanks!

thomaseizinger reviewed (commented):
Nice, thank you for that! A few more minor comments.

yellowred committed:
More references use relative location.

thomaseizinger commented on examples/rendezvous/Cargo.toml:
```diff
@@ -0,0 +1,14 @@
+[package]
+name = "rendezvous"
```
Can we consistently apply the naming of appending `-example` here? Thanks!

thomaseizinger commented on examples/rendezvous/src/bin/rzv-identify.rs:
```diff
@@ -19,48 +19,47 @@
-            .authenticate(libp2p_noise::NoiseAuthenticated::xx(&identity).unwrap())
-            .multiplex(libp2p_yamux::YamuxConfig::default())
+            .authenticate(libp2p::noise::NoiseAuthenticated::xx(&key_pair).unwrap())
+            .multiplex(libp2p::yamux::YamuxConfig::default())
```
Same with this topic here. I didn't explicitly comment on all lines but let's do this consistently across all examples and use `yamux::` etc in the code and `use libp2p::yamux` at the top. (For all `libp2p` modules: noise, yamux, etc)

yellowred replied:
sorry, missed it

thomaseizinger reviewed (commented):
Two more small requests, sorry for the back and forth!

yellowred committed:
Use relative path when importing.

yellowred committed:
cleanup

thomaseizinger reviewed (approved):
Thank you for tackling this, I am convinced this makes our repository a bit more approachable :)
